ganttproject/src/main/java/net/sourceforge/ganttproject/GanttProject.java

```diff
@@ -146,35 +148,35 @@ public class GanttProject extends GanttProjectBase implements ResourceView, Gant
 
   private FXSearchUi mySearchUi;
 
+  /**
+   * Holds the transaction ID specified by the server (String) and the local ID (Integer) of the last local transaction
+   * committed by the server.
+   * The local ID corresponds to the txn ID stored in the database.
+   */
   private static class TxnCommitInfo {
-    /** Base transaction ID received from the server. */
-    protected String baseTxnId = "";
-    /** Local ID of the last transaction committed by the server. */
-    protected int baseTxnLocalId = 0;
-
-    TxnCommitInfo() {}
-    TxnCommitInfo(String baseTxnId, int baseTxnLocalId) {
-      this.baseTxnId = baseTxnId;
-      this.baseTxnLocalId = baseTxnLocalId;
+    private final AtomicReference<ImmutablePair<String, Integer>> myTxnId;
+
+    TxnCommitInfo(String serverId, Integer localId) {
+      myTxnId = new AtomicReference<>(new ImmutablePair<>(serverId, localId));
     }
-  }
 
-  private static class TxnCommitInfoSync extends TxnCommitInfo {
-    synchronized Boolean compareAndSetBaseTxn(String expectedTxnId, String newTxnId) {
-      if (baseTxnId.equals(expectedTxnId)) {
-        baseTxnId = newTxnId;
-        baseTxnLocalId++;
-        return true;
-      }
-      return false;
+    /** If `oldTxnId` is currently being hold, sets the txn ID to `newTxnId` and moves the local ID ahead by `committedNum`. */
+    synchronized void update(String oldTxnId, String newTxnId, int committedNum) {
```

Comment:
I believe that with AtomicReference you don't need synchronized methods anymore. AtomicReference is thread-safe: https://jenkov.com/tutorials/java-util-concurrent/atomicreference.html

Reply:
Sure, I've missed it, sorry